timely_beliefs/sensors/func_store/knowledge_horizons.py

```diff
@@ -31,6 +32,45 @@ def at_date(
     return event_start - knowledge_time.astimezone(event_start.tzinfo)
 
 
+
+def x_years_ago_at_date(
+    event_start: datetime | pd.DatetimeIndex,
+    day: int,
+    month : int,
+    x : int = 1,
+    get_bounds: bool = False,
+    z: str | None = None,
+) -> timedelta | pd.TimedeltaIndex | tuple[timedelta, timedelta]:
+    """Compute the sensor's knowledge horizon to represent the event could be known since some date, `x` years ago.
+    
+    For example, it can be used for a tax rate that changes annually and with a known publication date. 
+
+    :param event_start:     Start of the event, used as an anchor for determining the knowledge horizon.
+    :param day:             Reference day of the month of the annual date to compare against.
+    :param month:           The month of the annual date to compare against.
+    :param x:               The number of years to shift the reference date to.
+    :param z:           Timezone string.
+    :param get_bounds:      If True, this function returns bounds on the possible return value.
+                            These bounds are normally useful for creating more efficient database queries when filtering by belief time.
+                            In this case, the knowledge horizon is unbounded.
+    """
+    if get_bounds:
+        return timedelta.min, timedelta.max
+    def x_years_ago_at_date_datetime(_event_start : datetime, day : int, month : int, x : int, z : str | None) -> timedelta:
+        if z is None:
+            z = _event_start.tzinfo
+        else:
+            z = timezone(z)
+
+        anchor = dict(year=_event_start.year - x, month=month, day=day, hour=0, minute=0, second=0, microsecond=0, tzinfo=z)
```

Comment:
Notes for myself:

I like the defaults you added here. This makes me dream of a generic function that allows setting all of these time components, with the relative components defaulting to 0.

Very much resembling the function signature of [relativedelta](https://dateutil.readthedocs.io/en/stable/relativedelta.html) (but then with support for timezone awareness).

Most of our knowledge horizon functions then become a simpler alias for a subset of this function.

```
def ex_ante_relativedelta(years=0, months=0, weeks=0, days=0, hours=0, year=None, month=None, day=None, weekday=None, hour=None):
    tz_naive_original_time = tz_aware_original_time.tz_convert(tz).tz_localize(None)
    tz_naive_earlier_time = (tz_naive_original_time - pd.Timedelta(years=years, months=months, weeks=weeks, days=days, hours=hours))
    if hour is not None:
        tz_naive_earlier_time = tz_naive_earlier_time.floor(
            "D"
        ) + pd.Timedelta(hours=hour)
    if minute is not None:
        tz_naive_earlier_time = tz_naive_earlier_time.floor(
            "H"
        ) + pd.Timedelta(minutes=minute)
    if second is not None:
        tz_naive_earlier_time = tz_naive_earlier_time.floor(
            "T"  # or "min"
        ) + pd.Timedelta(seconds=second)
    if microsecond is not None:
        tz_naive_earlier_time = tz_naive_earlier_time.floor(
            "s"
        ) + pd.Timedelta(microseconds=microsecond)
    tz_aware_earlier_time = tz_naive_earlier_time.tz_localize(tz).tz_convert(
        original_tz
    )

from functools import partial

x_days_at_y_o_clock = partial(ex_ante_relativedelta, days=x, hour=y)
x_years_ago_at_date = partial(ex_ante_relativedelta, years=x, month=month, day=day)
```

Likewise, `def ex_post_relativedelta(event_resolution, ...)` which should anchor the relativedelta to the event start rather than to the event end.

Reply:
This would be nice, indeed. Sadly, `Timedelta` [doesn't support years](https://pandas.pydata.org/docs/reference/api/pandas.Timedelta.html). Nevertheless, we can find an alternative way to handle this.